### ISSUE TITLE
Wish: plot activity of 'variables' 

### ISSUE THREAD

**Issue by feret** [Kappa-Dev/KaSim#471]
Let us say that I use the variable 'k' as the rate constant of several rules. 
It would be nice to have a way to plot the activity of 'k' that is to say the sum of the activity of all the rules the rate of which is defined as 'k'.

**Comment by glucksfall** [comment]
As far I know, it is possible to observe a variable. However, your request is more like "observe a rule" or "count how many time the rule fired at the time step"? Indeed it would be nice to have such feature in the syntax and don't rely on a token to count how many time a rule(s) fired at a given time.

best to the kappa-dev team

**Comment by feret** [comment]
Indeed, I am more interested in observing the cumulative activity of a pack of rules. 
I am dealing with a model of about 300 rules parametrized by only 12 rate constants. 
While it is easy to observe the activity of a rule by observing the number of embedding of its lhs in the state of the system, declaring the 12 observables are quite long to do by hand. 

Best.

**Comment by glucksfall** [comment]
Sure. I agree with you. I also wanted to measure the flow of the rules along with the agent quantities. Anything that reduces the amount of effort to write down code is beneficial.

By the way, how are you parameterizing the model? I'm asking because we're implementing a procedure to parameterize and I'm looking for bibliography and further discussion. If you prefer, I could contact you through e-mail.

Best!

**Comment by feret** [comment]
Unfortunately, I do it in a very artisanal way. 
We know that we would need at some point a procedure to assist model calibration, but we are kind of lacking man power to work in this direction. 

At this point, we add a model working with arbitrary rate constants. 
We get a lot of quantitative information which makes us change the whole model. 
I just want to make sure that the separation between time scales is not modified so much. For this, I compute pair-wise the ratio between each constant associated to a time-scale and a given arity (zeroary, unary, binary reactions). 

This is not ideal, but I do not see what to do with more appropriate tools.

**Comment by glucksfall** [comment]
I'm open to help and collaborate. We could talk later and formally with our groups about how to parameterize kappa models. And how to implement other useful analysis.

Best!

**Comment by feret** [comment]
Great!!! I am very interested. 
We keep in touch. 

Best.

**Comment by hmedina** [comment]
If you have the list of rule names, you could just parse the values out of
the influence map, no?


Back to the original question, should 'k' be more of an anchor, a pointer,
than just a group-by value?

A(),C() -> A(),. @ 'k'
B(),B() -> B(),. @ 'k'/2

The effective rate of the latter is just 'k', but the definition does not
say just 'k', so should they be tracked jointly? Should any complicated
algebraic expression that has the symbol 'k' be included, like other
variables, or only rules? What about reversible rules, and ambiguous rules,
how do you tally those? Some users might want to group by the forward model
(i.e. forward binary and forward unary), some might want to group by arity
(i.e. forward unary, and reverse rules), other might want to keep each
class separate...

Personally, I find using counters to be far cleaner. Specially with
MatLab's diff function, getting the activity becomes a one-liner:

A(),C() -> A(),. | f @ 'k' (0)
A(),C() -> A(),. | u @ 0 ('k'/100)
A(),. -> A(),C() | r @ 'k'
act = diff(f)/diff(time) + diff(u)/diff(time) + diff(r)/diff(time)

And adding a token to rules is quite easy, specially with regular
expressions; a little Python can help in going over the 300 rules, and
writing the analyzing script.

On Nov 22, 2017 11:13, "Jérôme Feret" <notifications@github.com> wrote:

Great!!! I am very interested.
We keep in touch.

Best.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Kappa-Dev/KaSim/issues/471#issuecomment-346398060>, or mute
the thread

**Comment by feret** [comment]
-> unary/binary/… 

My model is well typed. Constants for zeroary, unary, binary rates belong to three disjoint sets. 

-> Autos… 

Sure I consider rates up to multiplication by a constant. 

-> Tokens

By activity, I meant the speed of the system. 
Not the number of application of the rules. 
I would prefer avoid using an approximate encoding. 

Best.
